### PR TITLE
Fix issue #1167 (loading tiff ImageJ saved without units)

### LINKED ISSUE #1167: TIFF stack with no scale?
I've got a TIFF stack (256x256x103) with no scale set - this is what it looks like when I load it into FIJI:

gimp

Trying to load it with HyperSpy throws this error:

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-f56864702d12> in <module>()
----> 1 cs_si = hs.load('stack.tif')
      2 cs_si.change_dtype('float')
      3 X = cs_si.data.transpose(1,2,0)
      4 m, n, t = X.shape
      5 X = X.reshape(m*n, t)

/home/tjof2/hyperspy/hyperspy/io.py in load(filenames, signal_type, stack, stack_axis, new_axis_name, mmap, mmap_dir, **kwds)
    177             objects = [load_single_file(filename,
    178                                         **kwds)
--> 179                        for filename in filenames]
    180 
    181         if hyperspy.defaults_parser.preferences.Plot.plot_on_load:

/home/tjof2/hyperspy/hyperspy/io.py in <listcomp>(.0)
    177             objects = [load_single_file(filename,
    178                                         **kwds)
--> 179                        for filename in filenames]
    180 
    181         if hyperspy.defaults_parser.preferences.Plot.plot_on_load:

/home/tjof2/hyperspy/hyperspy/io.py in load_single_file(filename, signal_type, **kwds)
    221                                 reader=reader,
    222                                 signal_type=signal_type,
--> 223                                 **kwds)
    224 
    225 

/home/tjof2/hyperspy/hyperspy/io.py in load_with_reader(filename, reader, signal_type, **kwds)
    229                      **kwds):
    230     file_data_list = reader.file_reader(filename,
--> 231                                         **kwds)
    232     objects = []
    233 

/home/tjof2/hyperspy/hyperspy/io_plugins/tiff.py in file_reader(filename, record_by, **kwds)
    183                 _logger.info("Reading ImageJ tif metadata")
    184                 # ImageJ write the unit in the image description
--> 185                 units = image_description.split('unit=')[1].split('\n')[0]
    186                 scales = _get_scales_from_x_y_resolution(op)
    187 

IndexError: list index out of range
```

### PR DESCRIPTION
**Pull request by ericpre**
Fix #1167.
Fix issue for tiff files saved with imageJ without units and add better support for stacks. Add a few more test files and tests.

I could reproduce the bug and this PR should fix it.

### CODE REVIEW

**Comment by tjof2** [comment]
@ericpre  I'll test it myself tomorrow.

File: hyperspy/io_plugins/tiff.py [hunk 1]
```diff
@@ -181,9 +181,16 @@ def file_reader(filename, record_by='image', **kwds):
                 "Image_description tag: {0}".format(image_description))
             if 'ImageJ' in image_description:
                 _logger.info("Reading ImageJ tif metadata")
-                # ImageJ write the unit in the image description
-                units = image_description.split('unit=')[1].split('\n')[0]
-                scales = _get_scales_from_x_y_resolution(op)
+                try:
+                    # ImageJ write the unit in the image description
+                    if 'unit' in image_description:
+                        units = image_description.split('unit=')[1].split('\n')[0]
```
Comment (francisco-dlp) [inline hyperspy/io_plugins/tiff.py]:
```python
units = image_description.split('unit=')[1].strip()
``` 
maybe be more robust.

File: hyperspy/io_plugins/tiff.py [hunk 2]
```diff
@@ -181,9 +181,16 @@ def file_reader(filename, record_by='image', **kwds):
...
+                        if dc.ndim == 3 and not tiff.is_rgb:
+                            if 'spacing' in image_description:
+                                scales.append(float(image_description.split('spacing=')[1].split('\n')[0]))
```
Comment (francisco-dlp) [inline hyperspy/io_plugins/tiff.py]: As above: `split('\n')` -> `strip()`

**Comment by francisco-dlp** [comment]
This fixes the issue for ImageJ tif files but not for other tiff image stacks: the stack axis in missing when loading from tif. Example:

```python

>>> import numpy as np
>>> hs.signals.Signal2D(np.zeros((10, 20, 30))).save("test.tif")
>>> hs.load("test.tif")
<Signal2D, title: , dimensions: (|20, 10)>

```
Notice that

File: hyperspy/io_plugins/tiff.py (hunk 1 above)
Comment (ericpre) [inline hyperspy/io_plugins/tiff.py]: This is actually a split, there are several '\n' in the string.

**Comment by ericpre** [comment]
This should be better now. I tried to use the "axes names" of the tiff files to set the units and the scale to the right axes. I think this should be more robust.
I also put the unit/scale parsing code in a try statement, so it should skip it if an error is raised.

File: hyperspy/io_plugins/tiff.py [hunk 3]
```diff
@@ -274,86 +227,186 @@ def file_reader(filename, record_by='image', **kwds):
...
+        image_description = _decode_string(op["image_description"])
+        if "image_description_1" in op.keys():
+            image_description = _decode_string(op["image_description_1"])
+        _logger.debug(
+            "Image_description tag: {0}".format(image_description))
+        if 'ImageJ' in image_description:
+            _logger.debug("Reading ImageJ tif metadata")
+            # ImageJ write the unit in the image description
+            if 'unit' in image_description:
+                unit = image_description.split('unit=')[1].split('\n')[0]
```
Comment (francisco-dlp) [inline hyperspy/io_plugins/tiff.py]: As discussed before, `strip()` might be more robust than  `split('\n')[0]`

File: hyperspy/io_plugins/tiff.py [hunk 4]
```diff
@@ -274,86 +227,186 @@ def file_reader(filename, record_by='image', **kwds):
...
+                for key in ['x', 'y']: units[key] = unit
+                scales['x'], scales['y'] = _get_scales_from_x_y_resolution(op)
+            if 'spacing' in image_description:
+                scales['z'] = float(image_description.split('spacing=')[1].split('\n')[0])
```
Comment (francisco-dlp) [inline hyperspy/io_plugins/tiff.py]: As above.

**Comment by francisco-dlp** [comment]
Great, this fixes the issue. I'll merge it and I address my minor comment in a separate PR (we're planning to release v1.0.1 today).

**Comment by francisco-dlp** [comment]
@ericpre, I have just sent a PR with some minor fixes #1188, could you have a look at it?

**Pull request merged**